Bring up Llama 3.1 405B on two pods

tengyifei opened the pull request:
# What does this PR do?

These are the changes to make Llama 3.1 405B work on two Trillium TPU pods. It includes:

- Initialize the model layer by layer on the CPU, to workaround an OOM bug when initializing all layers in the model at once on the TPU.
- Save the `inv_freq` buffer in Llama for now, to ensure we can initialize it using `load_state_dict`. Otherwise, the `inv_freq` buffer will stay as a meta tensor.
- Adds a `USE_SINGLE_SLICE` env var before importing jax, to prevent jax from re-initializing MegaScale client. A corresponding custom libtpu build is required.
- Use a custom hybrid ring mesh to improve collectives performance.

tengyifei committed:
Add DCN sharding

tengyifei committed:
Suppress JAX megascale initialization

tengyifei committed:
Add log

tengyifei committed:
Monkey patch _create_device_mesh_for_nd_torus

tengyifei committed:
Add metrics before/after materializing model

tengyifei committed:
Materialize model layer by layer, to workaround OOM in multi-slice distributed model initialization

tengyifei committed:
Don't put inv_freq buffers in the state dict

See https://github.com/huggingface/transformers/issues/34234

tengyifei committed:
move model to XLA to be sure

tengyifei committed:
Attempt to support inv_freq another way

Instead of leaving it as a meta tensor, mark it as persistent=True

tengyifei committed:
Switch to a hybrid ring in 2 pod training

tengyifei committed:
Don't hardcode the number of slices

tengyifei committed:
Move around XLA metrics logs

tengyifei committed:
Replace nn.Linear with XlaPatchedLinear

zpcore commented on examples/pytorch/language-modeling/run_clm.py:
```diff
@@ -476,20 +478,58 @@ def main():
+    device_attributes = xr.global_runtime_device_attributes()
+    num_slices = max(int(d.get('slice_index', 0)) for d in device_attributes) + 1
+
+    if num_slices > 1 and model_args.spmd_2d_sharding == 0:
```
Should we give the flexibility to choose whether to use the 2d sharding or not?

zpcore commented on examples/pytorch/language-modeling/run_clm.py:
```diff
@@ -461,8 +461,10 @@ def main():
+        meta_device = torch.device('meta')
+        with meta_device:
             model = AutoModelForCausalLM.from_config(config, trust_remote_code=model_args.trust_remote_code)
+        # note: at this point, the mode is not materialized
```
NIT: typo `mode`.

zpcore commented:
This is great, can you point us to the material how does `hybrid ring mesh` work? Thanks

zpcore commented on examples/pytorch/language-modeling/run_clm.py:
```diff
@@ -476,20 +478,58 @@ def main():
+            # Multi-slice 2D sharding
+            tensor_axis = model_args.spmd_2d_sharding
+            fsdp_axis = num_devices // tensor_axis
+            mesh_shape = (fsdp_axis, tensor_axis)  # Should be (128, 4)
```
The comment seems only apply to 2 slices with 4 tensor parallelism.

zpcore commented on examples/pytorch/language-modeling/run_clm.py:
```diff
@@ -476,20 +478,58 @@ def main():
+            dcn_axis = num_slices
+            ici_fsdp_axis = num_devices // tensor_axis // dcn_axis
+            ici_mesh_shape = (ici_fsdp_axis, tensor_axis)
+            dcn_mesh_shape = (dcn_axis, 1)
```
Actually I don't understand what does `1` mean here, even though we are all using it :|

zpcore replied:
I think I got it

zpcore commented on examples/pytorch/language-modeling/run_clm.py:
```diff
@@ -507,7 +547,11 @@ def main():
+            dict_of_params[name] = param
 
+        model.load_state_dict(dict_of_params, assign=True)
+        model.to('xla')
```
Question: with meta device, did we skip loading the model into CPU RAM but loading into `XLA` device directly?

tengyifei committed:
Support remat every N layers (e.g. every other layer)

tengyifei committed:
[wip] integrate scan, offloading, AOTAutograd FA kernel

tengyifei committed:
works with scan

tengyifei committed:
tried activating offloading but it is mixing meta and xla devices

> /workspaces/torch/transformers/<eval_with_key>.6(210)forward()

206         mul_15 = torch.ops.aten.mul.Tensor(add_7, _to_copy_3);  add_7 = _to_copy_3 = None
207         sum_1 = torch.ops.aten.sum.dim_IntList(mul_15, [0, 1], True);  mul_15 = None
208         view_58 = torch.ops.aten.view.default(sum_1, [4096]);  sum_1 = None
209         _to_copy_4 = torch.ops.aten._to_copy.default(mul_14, dtype = torch.float32, layout = torch.strided, device = device(type='xla', index=0));  mul_14 = None
210  ->     mul_16 = torch.ops.aten.mul.Tensor(_to_copy_4, _to_copy_2)
211         mul_17 = torch.ops.aten.mul.Tensor(_to_copy_4, rsqrt_1);  _to_copy_4 = None
212         sum_2 = torch.ops.aten.sum.dim_IntList(mul_16, [2], True);  mul_16 = None
213         detach_29 = torch.ops.aten.detach.default(rsqrt_1);  rsqrt_1 = None
214         detach_30 = torch.ops.aten.detach.default(detach_29);  detach_29 = None

_to_copy_2 is `meta`.
_to_copy_4 is `xla`.

tengyifei committed:
use XLA tensors instead of meta tensors to trace the offloading graph

This supports hardcoded XLA device names in the input graph (e.g. due
to casts). We need to trace the graph correspondingly with XLA devices.

Why does a cast involve devices? Looks like the ATen ops doesn't include
cast, only `_to_copy`: https://pytorch.org/docs/stable/torch.compiler_ir.html

tengyifei committed:
Comment the testing script

tengyifei committed:
surface unexpected transfers to device when running the model

tengyifei committed:
surface transfers to device

tengyifei committed:
more trace spans

tengyifei committed:
move CurriedLayer out to __init__ so we don't leak a causal_mask on every step

tengyifei committed:
repro a particular llama3 8b training run

tengyifei committed:
support both 1d and 2d sharding with offload and scan

tengyifei committed:
fix NaN training loss in aot flash attention

tengyifei committed:
train 100 steps

tengyifei committed:
shard the logits

This helps XLA avoid gathering the logits and saves a few Gigs of RAM.

Before: http://shortn/_tlV88E1Ca3
After:  http://shortn/_GodQzu6GMu

tengyifei committed:
Get rid of aten_take in flash attention kernel

Using a custom aten slice fast path in the ptxla branch.

Before: http://shortn/_1Xyrm0lLdL
After:  http://shortn/_SgU6vG1pNm

tengyifei committed:
Test bs 16

After we trace out HLO in scan using placeholder tensors, memory usage
drops enough that we can up the batch size to 16.

Profile: http://shortn/_ESRGhAhKce

tengyifei committed:
Try removing sharding annotation in modeling_llama

These are negative optimizations (probably some are wrong).

On 2D sharded v6e-8, memory usage goes 26 GiB -> 22 GiB.
Step time: 4.76s -> 4.71s.

tengyifei committed:
copy the sharding from torchax llama405b scan version

v6e-8 llama8b 2d sharding HBM drops to 19.8 GiB. Let's see how it does
on 405b

Profile: http://shortn/_x9ZNp0KTh1

tengyifei committed:
Compute causal_mask inside the decoder layer

The 405b profile (http://shortn/_XRl3HxKpgn) looks very sus.
`bf16[126,4096,16384]` is gathered. One slice in the leading dimension
is updated, followed by a dynamic slice, then returned.

Remove one variable in the backward scan combine function for ease of
investigation.

tengyifei committed:
Add sharding annotation to attn_output

Looking at the graph in http://shortn/_n0fCn6xIt8, it appears the
tensors around attn_output/o_proj is unnecessarily all-gathered by XLA.
Let's see if this prevents that.

tengyifei committed:
Shard all backward scan outputs and shard o_proj weight gradients

From v6e-8 testing this seems to make the large FSDP all-gathering
of stacked o_proj weights go away. On Llama 3-8B, the stacked o_proj
is 32 layers * 4096 hidden * 4096 hidden / 2 TP * 2 bytes = 512 MiB.

Before:
  http://shortn/_bq5aqrTHdo
  HBM: 19898 MiB
  Step time: 4.705s

After:
  http://shortn/_2q04nF5faZ
  HBM: 19378 MiB
  Step time: 4.233s

tengyifei committed:
Attempt to diagnose MegaScale hang

tengyifei committed:
Wait for device ops at each step to debug MegaScale

tengyifei committed:
Revert "Wait for device ops at each step to debug MegaScale"

This reverts commit 41b61fef1396065d38e0c2cb56ce35f2dea26def.

tengyifei committed:
Disable jax megascale
